torch/_inductor/codegen/cpp_mha_template.py

```diff
@@ -0,0 +1,476 @@
+
+
+  // allocate per thread temp buf (accumulate type)
+  int64_t size_per_thread =
+        /* qk     */ qSplitSize * kvSplitSize +
+        /* qk_max */ qSplitSize +
+        /* qk_sum */ qSplitSize +
+        /* dst    */ qSplitSize * headSize;
+
+  {%- set acc_buf_name = "buf" %}
+      {{kernel.define_buffer(acc_buf_name, [num_thread, size_per_thread], dtype=accumulate_dtype)}}
+  {%- set acc_reduced_buf_name = "buf_reduced" %}
+      {{ kernel.define_buffer(acc_reduced_buf_name, [num_thread, qSplitSize, kvSplitSize], dtype=query_dtype)}}
+
+  const scalar_t* q_data = query;
+  const scalar_t* k_data = key;
+  const scalar_t* v_data = value;
+  auto  kv_indices_data = kv_indices;
+
+  scalar_t* out_data = output;
+
+  accum_t* buf_data = buf;
+  scalar_t* buf_reduced_data = is_reduced_type ? buf_reduced : nullptr;
+
+  at::parallel_for(0, batchSize * num_head * qSlice, 1, [&](int64_t begin, int64_t end) {
+     int64_t i = 0, j = 0, k = 0;
+     at::native::data_index_init(begin, i, batchSize, j, num_head, k, qSlice);
+        int ompIdx = at::get_thread_num();
+        accum_t* buf_ptr = buf_data + ompIdx * size_per_thread;
+        accum_t* qk_data = buf_ptr;
+        accum_t* qk_max_data = qk_data + qSplitSize * kvSplitSize;
+        accum_t* qk_sum_data = qk_max_data + qSplitSize;
+        accum_t* dst_data = qk_sum_data + qSplitSize;
+        scalar_t* qk_reduced_data = is_reduced_type ? buf_reduced_data + ompIdx * qSplitSize * kvSplitSize : nullptr;
+        scalar_t* query_t_padding_ptr = nullptr;
+
+        for ([[maybe_unused]] auto z : c10::irange(begin, end)) {
+          int64_t m = k * qSplitSize;
+          int64_t cur_qSplitSize = std::min(qSplitSize, qSize - m);
+          // Initialize max and sum
+          fill_stub(qk_max_data,
+              -std::numeric_limits<accum_t>::infinity(), cur_qSplitSize);
+          fill_stub(qk_sum_data,
+              static_cast<accum_t>(0), cur_qSplitSize);
+          int64_t num_keys = kvSize;
+
+          for (int64_t n = 0; n < num_keys; n += kvSplitSize) {
+            int64_t cur_kvSplitSize = std::min(kvSplitSize, kvSize - n);
+            cur_kvSplitSize = cur_kvSplitSize == kvSplitSize ? kvSplitSize : kvTail;
+
+            // Calculate scale * q @ k.T
+            auto kv_block_num = n / kvBlockSize;
+            auto kv_block_offset = n - kv_block_num * kvBlockSize;
+            // getting kv indices by [BS, Head, 1, kv_block_num]
+            auto i_kvi = is_broadcast_bs_kvi ? i/bs_shards_kvi : i;
+            auto j_kvi = is_broadcast_head_kvi ? j/gqa_shards_kvi : j;
+            auto kv_logical_data = kv_indices_data + i_kvi*kviStrideB + j_kvi*kviStrideH  + kv_block_num;
+            auto i_kv = is_broadcast_bs_kv ? i/bs_shards : i;
+            auto j_kv = is_broadcast_head_kv ? j/gqa_shards : j;
+            auto k_addr = k_data + i_kv * kStrideB + j_kv * kStrideH + (*kv_logical_data * kvBlockSize)  * kStrideN;
+
+            at::native::cpublas::gemm(
```

Comment:
We'd better call cpp micro-kernel gemm template instead. This is because:
1. The micro-kernel gemm template brings better perf with shape specializations and also has less overhead than `cpublas::gemm` which might have to pay primitive creation and JIT compilation runtime overhead.
2. `cpublas::gemm` is C++ API. Invoking it from the inductor codegen breaks the ABI compatibility requirement of AOT inductor.

Reply:
Got it, we will look into this.

Reply:
Here have refined to use `cpublas::brgemm`  as discussed, and using micro-kernel gemm template  and adding runtime weight pack will be the next step.